modules/caddyhttp/reverseproxy/reverseproxy.go

```diff
@@ -922,7 +922,9 @@ func (h *Handler) finalizeResponse(
 ) error {
 	// deal with 101 Switching Protocols responses: (WebSocket, h2c, etc)
 	if res.StatusCode == http.StatusSwitchingProtocols {
-		h.handleUpgradeResponse(logger, rw, req, res)
+		var wg sync.WaitGroup
+		h.handleUpgradeResponse(logger, &wg, rw, req, res)
+		wg.Wait()
```

Comment:
This blocks -- not sure if that's actually a good idea?

Reply:
This adds a little bit of delay so that statistics are properly updated. Both ends of the connection are closed when `handleUpgradeResponse` returns. If this blocks, go has a serious issue where `net.Conn.Close` doesn't unblock `Read` or `Write`.

Reply:
Ah, so you're saying that `h.handleUpgradeResponse()` was already blocking, and `wg.Wait()` waits until the actual conns are `Close()`ed, but otherwise it's basically the same amount of blocking.

In that case, this LGTM.

I just couldn't remember if this goroutine was returning while the hijacked connection kept running to do its copy, or if we blocked in there, but it looks like handleUpgradeResponse has a select that blocks at least for most of the time.

So yeah, if this patch works, and doesn't block much longer, then I think we can merge it.

Thanks!!

Reply:
To clarify, does the size still log as 0, with this patch?

Reply:
Before this patch, sometimes size logs as 0. After this, none does.